Migrate to canCurrentProcessAccessPasskeysForRelyingParty

pascoej opened the pull request:
#### 6080b48057a6b5c1047e5c0e25b343e6160653f7
<pre>
Migrate to canCurrentProcessAccessPasskeysForRelyingParty
<a href="https://rdar.apple.com/119825781">rdar://119825781</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=267442">https://bugs.webkit.org/show_bug.cgi?id=267442</a>

Reviewed by Brent Fulgham.

There are some sync methods used that now have async equivalents. We should move to using
them. These calls can be slow when the daemon has to start up.

* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:

Canonical link: <a href="https://commits.webkit.org/273098@main">https://commits.webkit.org/273098@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3d112cc4790c557f5ed5399fb160e74f0885ecb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34215 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36204 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36904 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31000 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15444 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10154 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30032 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34727 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11057 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30533 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9624 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9758 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38195 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31121 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30884 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35819 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9856 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7723 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33738 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11642 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7885 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10415 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10674 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
```diff
@@ -30,7 +30,9 @@
 @property (class, nonatomic) BOOL shouldUseAlternateCredentialStore;
 + (BOOL)arePasskeysDisallowedForRelyingParty:(nonnull NSString *)relyingParty;
++ (void)getArePasskeysDisallowedForRelyingParty:(nonnull NSString *)relyingParty withCompletionHandler:(void (^ _Nonnull)(BOOL))completionHandler NS_REFINED_FOR_SWIFT;
 + (BOOL)canCurrentProcessAccessPasskeysForRelyingParty:(nonnull NSString *)relyingParty;
```
Why do you leave a declaration for `canCurrentProcessAccessPasskeysForRelyingParty` in the header, when you seem to remove it in the implementation file?

pascoej replied:
I thought `getArePasskeysDisallowedForRelyingParty` and it's sync equivalent were introduced at different times, but it turns out that's not true. I've removed usage of `canCurrentProcessAccessPasskeysForRelyingParty` now.

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -846,22 +846,43 @@ static inline void continueAfterRequest(RetainPtr<id <ASCCredentialProtocol>> cr
-        return [getASCWebKitSPISupportClass() canCurrentProcessAccessPasskeysForRelyingParty:data.securityOrigin()->domain()];
-    return false;
+        handler([getASCWebKitSPISupportClass() canCurrentProcessAccessPasskeysForRelyingParty:data.securityOrigin()->domain()]);
+    handler(false);
```
This means we always call `handler(false)`, even if we just invoked the handler with the result of `canCurrentProcessAccessPasskeysForRelyingParty`. Is that right? Seems wrong to invoke the completion handler twice in.a row.

pascoej replied:
Fixed.

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -846,22 +846,43 @@ static inline void continueAfterRequest(RetainPtr<id <ASCCredentialProtocol>> cr
         return;
     }
+    if ([getASCWebKitSPISupportClass() respondsToSelector:@selector(arePasskeysDisallowedForRelyingParty:)])
+        handler([getASCWebKitSPISupportClass() arePasskeysDisallowedForRelyingParty:data.securityOrigin()->domain()]);
```
Again, we first invoke the handler with the "real" answer, then always call with `false`. That seems wrong.

pascoej replied:
Fixed.

brentfulgham reviewed (changes requested):
Can you double-check your handler logic? I think you are potentially calling twice in a. row with the wrong answer.

brentfulgham reviewed (approved):
Looks great! r=me

pascoej committed:
Migrate to canCurrentProcessAccessPasskeysForRelyingParty
rdar://119825781
https://bugs.webkit.org/show_bug.cgi?id=267442

Reviewed by Brent Fulgham.

There are some sync methods used that now have async equivalents. We should move to using
them. These calls can be slow when the daemon has to start up.

* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:

Canonical link: https://commits.webkit.org/273098@main
